pydantic/_internal/_fields.py

```diff
@@ -53,6 +56,60 @@ def get_type_hints_infer_globalns(
     return get_type_hints(obj, globalns=globalns, localns=localns, include_extras=include_extras)
 
 
+def get_typed_dict_type_hints(
+    typed_dict_cls: Any,
+    types_namespace: dict[str, Any] | None = None,
+) -> dict[str, Any]:
+    """Collect the type hints of a TypedDict.
+
+    This takes into account any TypeVar transformations when using generic
+    TypedDict.
+
+    Args:
+        typed_dict_cls: a TypedDict subclass
+        types_namespace: Optional extra namespace to look for types in.
+
+    Returns:
+        the final type hints.
+    """
+    hints: dict[str, Any] = {}
+    typevars_map: dict[type, dict[Any, Any]] = {}
+    required_keys: frozenset[str] = typed_dict_cls.__required_keys__
+    previous_base_cls: type | None = None
+
+    for base_cls in reversed(mro(typed_dict_cls)):
+        base = base_cls
+        if base in typed_dict_types:
+            continue
+
+        # needed if base_cls is a generic alias
+        if get_args(base_cls):
+            base: Any = get_origin(base_cls)
+
+        new_type_hints = get_cls_type_hints_lenient(base, types_namespace)
```

Comment:
This sort of looks like it's an `O(n^2)` algorithm, because this function iterates `base.__mro__` too.

I wonder whether there's an adjusted form of this function which iterates the whole mro once, collecting the `typevars_map` and also a mapping of `{ field_name -> (annotation, base) }`. Then that mapping can be transformed into the `hints` dictionary at the end of the iteration.

Reply:
This could potentially be `base.__annotations__` instead of `get_cls_type_hints_lenient` on second thought